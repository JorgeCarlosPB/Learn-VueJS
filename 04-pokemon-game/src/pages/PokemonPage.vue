<template>
  <div>
    <h1 v-if="!pokemon">Espere por favor...</h1>
    <div v-else>
        <h1>¿Quién es este Pokemón?</h1>

        <!--TODO: imgagen Pokemón-->
        <Pokemon-picture 
            :pokemonId="pokemon.id" 
            :showPokemon="showPokemon" 
        />

        <!--ToDO: 4 Opciones-->
        <Pokemon-options  
            :pokemons="pokemonArr"
            @selection="checkAnswer"
        />

        <template v-if="showAnswer">
            <h2  class="fade-in">{{message}}</h2>
            <button @click="newGame">Nuevo juego</button>
        </template>

        
    </div>


  </div>
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions'
import PokemonPicture from '@/components/PokemonPicture'
import getPokemonOptions from   '@/helpers/getPokemonOptions'



export default {
    components: {
        PokemonOptions,
        PokemonPicture  
    },
    data(){
        return{
            pokemonArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods: {
        async mixPokemonArray(){
            this.pokemonArr = await getPokemonOptions()

            const rdnInt = Math.floor(Math.random()*4)
            this.pokemon = this.pokemonArr[rdnInt]
        },
        checkAnswer(selectedID){
            this.showPokemon = true
            this.showAnswer = true

            if(selectedID === this.pokemon.id){
                this.message = `Correcto, ${this.pokemon.name}`
            }else{
                this.message = `Oops, era ${this.pokemon.name}`
            }
            
        },
        newGame(){
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonArr = []
            this.pokemon = null
            this.mixPokemonArray()
        }
    },
    mounted(){
        this.mixPokemonArray()
    }

}
</script>

<style>
    

</style>