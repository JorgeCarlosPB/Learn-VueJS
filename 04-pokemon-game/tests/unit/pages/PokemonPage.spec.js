import { shallowMount } from "@vue/test-utils";
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from "../mocks/pokemons.mock";

describe('PokemonPage component', ()=>{
    let wrapper
    

    beforeEach(()=>{
        wrapper=shallowMount(PokemonPage)
    })

    test('Debe de hacer match con el snapshot', ()=>{
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de llamar el mixPokemonArray al montar',()=>{
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        const wrapper = shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })

    test('Debe de hacer match con el snapshot cuando cargan los pokemons',()=>{
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe de mostrat los componentes de PokemonPicture y PokemonOptions',()=>{
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')

        
        expect(picture.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy()
        //PokemonPicture debe de existir
        //PokemonOptions debe de existir

        expect(picture.attributes('pokemonid')).toBe('1')
        expect(options.attributes('pokemons')).toBeTruthy()
        //PokemonPicture atribute pokemon id===5
        //PokemonOptionsatribute pokemons toBe true
    })

    test('Pruebas con chekanswer',async()=>{
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        //console.log(wrapper.vm)

        await wrapper.vm.checkAnswer(1)
        expect(wrapper.find('h2').exists()).toBeTruthy()
        
        //console.log(wrapper.find('h2').text())
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(wrapper.find(h2).text()).toBe(`Correcto, ${pokemons[0].name}`)

        await wrapper.vm.checkAnswer(2)
        expect(wrapper.vm.message).toBe(`Oops, era ${this.pokemon.name}`)
        




    })
})