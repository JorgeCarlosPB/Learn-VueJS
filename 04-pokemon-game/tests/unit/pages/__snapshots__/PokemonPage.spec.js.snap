// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>Espere por favor...</h1>
</div>
`;

exports[`PokemonPage component Debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <div>
    <h1>¿Quién es este Pokemón?</h1>
    <!--TODO: imgagen Pokemón-->
    <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
    <!--ToDO: 4 Opciones-->
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <!--v-if-->
  </div>
</div>
`;
